application/common/models/User.php

```diff
@@ -40,6 +48,195 @@ public function rules()
+
+        $user = self::findOne(['employee_id' => $personnelUser->employeeId]);
+        if ( ! $user) {
+            $user = new User();
+            $user->employee_id = $personnelUser->employeeId;
+            $user->first_name = $personnelUser->firstName;
+            $user->last_name = $personnelUser->lastName;
+            $user->idp_username = $personnelUser->username;
+            $user->email = $personnelUser->email;
+            if ( ! $user->save()) {
+                /**
+                 * @todo add logging with model validation errors
+                 */
+                throw new \Exception('Unable to create new user', 1456760294);
+            }
+        } else {
+            $user->updateProfileIfNeeded(
+                $personnelUser->firstName,
+                $personnelUser->lastName,
+                $personnelUser->username,
+                $personnelUser->email
+            );
+        }
+
+        return $user;
+    }
+
+
+    /**
+     * Update local user record if given properties are different than currently stored
+     * @param string $firstName
+     * @param string $lastName
+     * @param string $username
+     * @param string $email
+     * @return bool True if profile was updated, false if not
+     * @throws \Exception
+     */
+    public function updateProfileIfNeeded($firstName, $lastName, $username, $email)
+    {
+        $dirty = false;
+        $properties = [
+            'first_name' => $firstName,
+            'last_name' => $lastName,
+            'idp_username' => $username,
+            'email' => $email,
+        ];
+
+        foreach ($properties as $property => $value) {
+            if ($this->$property != $value) {
+                $dirty = true;
+                $this->$property = $value;
+            }
+        }
+
+        if ($dirty) {
+            if ($this->save()) {
+                return true;
+            } else {
+                /**
+                 * @todo add logging with model validation errors
+                 */
+                throw new \Exception('Unable to update profile', 1456760819);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Return array of arrays of masked out methods
+     * @return array
+     */
+    public function getMaskedMethods()
+    {
+        $methods = [];
+        foreach ($this->methods as $method) {
+            $methods[] = [
+                'uid' => $method->uid,
+                'type' => $method->type,
+                'value' => $method->getMaskedValue(),
+            ];
+        }
+        return $methods;
+    }
+
+    /**
+     * @return PersonnelUser
+     * @throws \Exception
+     */
+    public function getPersonnelUser()
+    {
+        if ( ! empty($this->personnelUser)) {
+            return $this->personnelUser;
+        }
+
+        /*
+         * If session is available, cache data in session
+         */
+        try {
+            $sessionAvailable = (\Yii::$app->user->identity == $this);
```

Comment:
Is this logic something that should be pulled out into a separate function? It seems non-intuitive enough that it might be worth having some unit tests to prove that it does what it says. Doing so could also improve the readability of this code.